File: Calculator/calc.cpp
#include <iostream>
using namespace std;

#include "calc.h"
#include <cstring>

Calc::Calc(int argcIn, char* argvIn[])
{
  inFix = new char[strlen(argvIn[1])+1];
  strcpy(inFix, argvIn[1]);
  if (!CheckTokens(argvIn[1]) || !CheckParens(argvIn[1]))
  {
    cout << "Invalid expression detected" << endl;
    exit(EXIT_FAILURE);
  }
  
  //Make values 
  values = new int[26];
  for (int o = 0; o < 26; o++)
    values[o] = 0;
  //Fill values
  int numOps = argcIn;
  for (int i = 2; i < numOps; i++)
     values[i-2] = atoi(argvIn[i]);

  InFixToPostFix();
}

Calc::~Calc()
{
  delete values;
  delete inFix;
  delete postFix;
}

bool Calc::CheckTokens(char* tempArr)
{
  
  for (int i = 0; tempArr[i] != '\0'; i++)
  {
    if(tempArr[i] != '(' && tempArr[i] != ')' && !isdigit(tempArr[i]) && !isalpha(tempArr[i]) && int(tempArr[i]) != 42 && int(tempArr[i]) != 43 && int(tempArr[i]) != 45 && int(tempArr[i]) != 47)
    {
      return false;
      cout << "Bad Tokens" << endl;
    }
  }

  return true;
}

bool Calc::CheckParens(char* tempArr)
{
  int openTally = 0;
  int closeTally = 0;
  for (int i = 0; tempArr[i] != '\0'; i++)
  {
    if (tempArr[i] == '(')
      openTally++;
    if (tempArr[i] == ')')
      closeTally++;
  }

  if (openTally == closeTally)
   return true;
  else
   return false;
}

void Calc::InFixToPostFix()
{
  //Make space
  postFix = new char[strlen(inFix)+1];
  Stack* stk = new Stack; 
  int p = 0;
  //Main loop
  for (int k = 0; inFix[k] != '\0'; k++)
  {
    if (isalpha(inFix[k]))
    {
      postFix[p] = inFix[k];
      p++;
    }
    if (inFix[k] == '(')
      stk->Push(inFix[k]);
    if (int(inFix[k]) == 42 || int(inFix[k]) == 43 || int(inFix[k]) == 45 || int(inFix[k]) == 47 )
      stk->Push(inFix[k]);
    if (inFix[k] == ')')
    {
      while (stk->Peek() != '(')
      {
        postFix[p] = stk->Peek();
        p++;
        stk->Pop();
      }
      stk->Pop();
    }

  }
}

int Calc::Evaluate()
{
  Stack* stk = new Stack; 
  int result = 0;
  for (int k = 0; postFix[k] != '\0'; k++)
  {
    if (isalpha(postFix[k]))
      stk->Push(postFix[k]);
    else
    {
      char op2 = stk->Peek();
      stk->Pop();
      char op1 = stk->Peek();
      stk->Pop();
      //Multiplication
      if (int(postFix[k]) == 42)
        result = values[op1 - 'A'] * values[op2 - 'A'];
      //Addition
      if (int(postFix[k]) == 43)
        result = values[op1 - 'A'] + values[op2 - 'A'];
      //Subtraction
      if (int(postFix[k]) == 45)
        result = values[op1 - 'A'] - values[op2 - 'A'];
      //Division
      if (int(postFix[k]) == 47)
        result = values[op1 - 'A'] / values[op2 - 'A'];
    }
 }
 return result;
}

void Calc::DisplayInFix()
{
 cout << "inFix: " << endl;
  for (int k = 0; inFix[k] != '\0'; k++)
    cout << inFix[k];
  cout << endl;
}

void Calc::DisplayPostFix()
{
 cout << "postFix: " << endl;
 for (int k = 0; postFix[k] != '\0'; k++)
   cout << postFix[k];
 cout << endl;
}

File: Calculator/calc.m
#include <iostream>
using namespace std;

#include "calc.h"
#include <cstring>


Calc::Calc(char* argvIn)
{
  inFix = new char[strlen(argvIn)+1];
  strcpy(inFix, argvIn);
 
  if(CheckTokens() == false)
  {
    cout << "Invalid character" << endl;
    exit(EXIT_FAILURE);
  }
  MakeValueTbl();
  Parse();
  CheckParens();
  DisplayInFix();
}

Calc::~Calc()
{
  delete []inFix;
}

bool Calc::CheckTokens()
{
  string eq = inFix;
  int i = 0;
  bool returnVal = false;
  while(eq[i] != '\0')
  {
    if(eq[i]  == '*' || eq[i] == '/' || eq[i] == '+' || eq[i] == '-')
      returnVal = true;
  
    if(eq[i] <= 48 && eq[i] >= 57)
      returnVal = true;

    if(eq[i] == '(' || eq[i] == ')')
      returnVal == true;
    i++;
  }
 return returnVal;
}

void Calc::MakeValueTbl()
{
  int arr[26];
  valueTbl = arr;
  for(int i = 0; i < 26; i++)
    valueTbl[i] = 0;
}


void Calc::Parse()
{

  char* ptr;
  char* tmp = new char[strlen(inFix) + 1];
  char* str = new char[strlen(inFix) + 1];
  strcpy(str, inFix);
  int* tokens = new int[26];
  char delimiters[] = " ()+_*/";

  //turn char into digits
  int i = 0;
  ptr = strtok(str,delimiters);
  while (ptr != NULL)
  {
    tokens[i] = atoi(ptr); 
    ptr = strtok(NULL,delimiters);
    i++;
  }
  
  //fill valueTbl
  for (int j = 0; j < i; j++)
    valueTbl[j] = tokens[j];
   
  i = 0;
  int x = 0; 
  valueIdx = 0;
  
  while(inFix[i] != '\0')
  {
    if(inFix[i] == '(' || inFix[i] == ')')
      tmp[x] = inFix[i];
    else if(inFix[i]  == '*' || inFix[i] == '/' || inFix[i] == '+' || inFix[i] == '-')
      tmp[x] = inFix[i];
    else
      {
        tmp[x] = valueIdx + 65;
        i = FindLast(i) - 1;
        valueIdx++;
      }
    i++;
    x++;   
  } 
  
  delete []inFix;
  inFix = tmp;
  tmp = NULL;
}

bool Calc::CheckParens()
{
  Stack *stk = new Stack;
  int i = 0;
  char ch = inFix[i];
  bool bal = true;
  
  while(inFix[i] != '\0')
  {
    if(ch == '(')
      stk->Push(ch);
    else if(ch == ')')
      stk->Push(ch);
    else
    {
     if (!stk->IsEmpty())
       stk->Pop();
     else
       bal = false; 
    }
    i++;
  }
  if(bal && stk->IsEmpty())
    return true;
  return false;
}


int Calc::FindLast(int cur)
{
  char ch = inFix[cur];
  while(ch >= '0' && ch <= '9')
  {
    cur++;
    ch = inFix[cur];
  }
  return cur;
}

void Calc::DisplayInFix()
{
  for(int i = 0; i < strlen(inFix)+1; i++)
    cout << inFix[i];
}
